### PR TITLE
Temporary fix to speed up `canAvoidInstr`

### PR DESCRIPTION
**Pull request by DiligentPenguinn**
Following the fact that the check `canAvoidInstr` implemented in previous PR #1687 is too slow, this PR implements a temporary fix by caching the result of `isEnvDependent` as a field for each `ControlItem`

### CODE REVIEW

**Review by RichDom2185 — COMMENTED** [review]
Looks good for a hotfix, thanks!

Though I'd remove `isEnvDependent` entirely and move the logic to the node creator function instead. What do you think? Do you foresee any more changes to this logic? You mentioned this is just a basic check (for the demo the other day) right? What future more "advanced" checks might there be?

Also, can I check what's left of this PR before it can be marked as ready?

Thanks!

**Comment by DiligentPenguinn** [comment]
@RichDom2185 Yes, I think that moving the logic to the node creator function is what should ultimately be done, but I won't be able to work on it very soon (due to exams). Until then I think it is still considerably slow with this fix though so I'm trying other ways to speed it up a bit more first. If you think it's fine for now then I'll mark it as ready

**Comment by DiligentPenguinn** [comment]
@RichDom2185 I just pushed approach 2, this seems to fix the speed issue, and I think it's closer to what would be done when we move the logic to the node creator function eventually.

File: src/cse-machine/interpreter.ts [hunk 1]
```diff
@@ -98,15 +98,34 @@ type CmdEvaluator = (
...
   public constructor(program?: es.Program | StatementSequence) {
     super()
-
+    this.numEnvDependentItems = 0
     // Load program into control stack
     program ? this.push(program) : null
   }
 
+  public canAvoidEnvInstr(): boolean {
+    return this.numEnvDependentItems == 0
```
Comment (RichDom2185) [inline src/cse-machine/interpreter.ts:110]: Let's use strict equality here

```suggestion
    return this.numEnvDependentItems === 0
```

File: src/cse-machine/interpreter.ts [hunk 2]
```diff
@@ -98,15 +98,34 @@ type CmdEvaluator = (
...
+  }
+
+  public pop(): ControlItem | undefined {
+    const item = super.pop()
+    if (item != undefined && isEnvDependent(item)) {
```
Comment (RichDom2185) [inline src/cse-machine/interpreter.ts:115]: Ditto.

```suggestion
    if (item !== undefined && isEnvDependent(item)) {
```

File: src/cse-machine/interpreter.ts [hunk 3]
```diff
@@ -98,15 +98,34 @@ type CmdEvaluator = (
...
+      this.numEnvDependentItems--
+    }
+    return item
+  }
+
   public push(...items: ControlItem[]): void {
     const itemsNew: ControlItem[] = Control.simplifyBlocksWithoutDeclarations(...items)
+    // testing
```
Comment (RichDom2185) [inline src/cse-machine/interpreter.ts:123]: Can we remove this comment?

File: src/cse-machine/utils.ts
Comment (RichDom2185) [inline src/cse-machine/utils.ts:1]: I'm good with this as long as it is a temporary hotfix. Please create an issue to migrate this to the node creator in the future.

Thanks!

**Review by RichDom2185 — COMMENTED** [review]
Just some minor nits, otherwise all good, thanks a lot!

File: src/cse-machine/interpreter.ts (hunk 3 above)
Comment (RichDom2185) [reply src/cse-machine/interpreter.ts]: Resolved in eca7bbf

File: src/cse-machine/interpreter.ts (hunk 2 above)
Comment (RichDom2185) [reply src/cse-machine/interpreter.ts]: Resolved in eca7bbf

File: src/cse-machine/interpreter.ts (hunk 1 above)
Comment (RichDom2185) [reply src/cse-machine/interpreter.ts]: Resolved in eca7bbf

File: src/cse-machine/utils.ts
Comment (RichDom2185) [reply src/cse-machine/utils.ts:1]: See #1702.

**Comment by RichDom2185** [comment]
Part of #1698 (does not close it fully due to the comment above).

**Pull request merged**